parse/planner/planner.go

```diff
@@ -0,0 +1,2392 @@
+		}
+
+		// rewrite the having expression to use the aggregate functions
+		havingExpr, err = s.rewriteAccordingToAggregate(havingExpr, groupingTerms, aggTerms)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+
+		plan = &Filter{
+			Child:     plan,
+			Condition: havingExpr,
+		}
+	}
+
+	// now we need to rewrite the select list to use the aggregate functions
+	for i, resultCol := range results {
+		results[i].Expr, err = s.rewriteAccordingToAggregate(resultCol.Expr, groupingTerms, aggTerms)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+	}
+
+	// finally, all of the aggregated columns need to be added to the Aggregate node
+	for _, agg := range order.OrderMap(aggTerms) {
+		aggPlan.AggregateExpressions = append(aggPlan.AggregateExpressions, agg.Value.Expr)
+		aggregateRel.Fields = append(aggregateRel.Fields, agg.Value.Field)
+	}
+
+	return plan, aggregateRel, func(lp LogicalPlan) (LogicalPlan, *Relation) {
+		var resultColExprs []LogicalExpr
+		var resultFields []*Field
+		for _, resultCol := range results {
+			resultColExprs = append(resultColExprs, resultCol.Expr)
+			resultFields = append(resultFields, resultCol.Field)
+		}
+
+		var p LogicalPlan = &Project{
+			Child:       lp,
+			Expressions: resultColExprs,
+		}
+
+		if node.Distinct {
+			p = &Distinct{
+				Child: p,
+			}
+		}
+
+		return p, &Relation{
+			Fields: resultFields,
+		}
+	}, nil
+}
+
+// hasAggregate returns true if the expression contains an aggregate function.
+func hasAggregate(expr LogicalNode) bool {
+	var hasAggregate bool
+	traverse(expr, func(node Traversable) bool {
+		if _, ok := node.(*AggregateFunctionCall); ok {
+			hasAggregate = true
+			return false
+		}
+
+		return true
+	})
+
+	return hasAggregate
+}
+
+// exprFieldPair is a helper struct that pairs an expression with a field.
+// It uses a generic because there are some times where we want to guarantee
+// that the expression is an IdentifiedExpr, and other times where we don't
+// care about the concrete type.
+type exprFieldPair[T LogicalExpr] struct {
+	Expr  T
+	Field *Field
+}
+
+// rewriteAccordingToAggregate rewrites an expression according to the rules of aggregation.
+// This is used to rewrite both the select list and having clause to validate that all columns
+// are either captured in aggregates or have an exactly matching expression in the group by.
+func (s *scopeContext) rewriteAccordingToAggregate(expr LogicalExpr, groupingTerms map[string]*IdentifiedExpr, aggTerms map[string]*exprFieldPair[*IdentifiedExpr]) (LogicalExpr, error) {
+	node, err := Rewrite(expr, &RewriteConfig{
+		ExprCallback: func(le LogicalExpr) (LogicalExpr, bool, error) {
+			// if it matches any group by term, we need to rewrite it
+			// and stop traversing any children
+			identified, ok := groupingTerms[le.String()]
+			if ok {
+				return &ExprRef{
+					Identified: identified,
+				}, false, nil
+			}
+
+			switch le := le.(type) {
+			case *ColumnRef:
```

Comment:
should also check AliasExpr?

Reply:
An example is `SELECT SUM(a) as sa FROM table1 GROUP BY a HAVING sa > 10`

Reply:
That's not a valid query:

```
postgres=# select sum(age) as aa from users group by name having aa > 10;
ERROR:  column "aa" does not exist
LINE 1: ...lect sum(age) as aa from users group by name having aa > 10;
```

Reply:
You have a point though, as this is a valid query:

```
select sum(age) as aa from users group by name order by aa;
```

Reply:
> That's not a valid query:
> 
> ```
> postgres=# select sum(age) as aa from users group by name having aa > 10;
> ERROR:  column "aa" does not exist
> LINE 1: ...lect sum(age) as aa from users group by name having aa > 10;
> ```

hmmm, I was testing using SQLite.. :(

Reply:
Ok, I think the solution is to project before sort, but we will make sure every column is projected. We can then perform projection pushdown to optimize this.